action/protocol/staking/candidate_statemanager.go

```diff
@@ -136,18 +166,158 @@ func (csm *candSM) CreditBucketPool(amount *big.Int) error {
 }
 
 func (csm *candSM) DebitBucketPool(amount *big.Int, newBucket bool) error {
-	return csm.bucketPool.DebitPool(csm.StateManager, amount, newBucket)
+	return csm.bucketPool.DebitPool(csm, amount, newBucket)
 }
 
 func (csm *candSM) Commit() error {
 	if err := csm.candCenter.Commit(); err != nil {
 		return err
 	}
 
-	if err := csm.bucketPool.Commit(csm.StateManager); err != nil {
+	if err := csm.bucketPool.Commit(csm); err != nil {
 		return err
 	}
 
 	// write updated view back to state factory
 	return csm.WriteView(_protocolID, csm.DirtyView())
 }
+
+func (csm *candSM) getBucket(index uint64) (*VoteBucket, error) {
+	return srToCsr(csm).getBucket(index)
+}
+
+func (csm *candSM) updateBucket(index uint64, bucket *VoteBucket) error {
+	if _, err := csm.getBucket(index); err != nil {
+		return err
+	}
+
+	_, err := csm.PutState(
+		bucket,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(bucketKey(index)))
+	return err
+}
+
+func (csm *candSM) putBucket(bucket *VoteBucket) (uint64, error) {
+	var tc totalBucketCount
+	if _, err := csm.State(
+		&tc,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(TotalBucketKey)); err != nil && errors.Cause(err) != state.ErrStateNotExist {
+		return 0, err
+	}
+
+	index := tc.Count()
+	// Add index inside bucket
+	bucket.Index = index
+	if _, err := csm.PutState(
+		bucket,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(bucketKey(index))); err != nil {
+		return 0, err
+	}
+	tc.count++
+	_, err := csm.PutState(
+		&tc,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(TotalBucketKey))
+	return index, err
+}
+
+func (csm *candSM) delBucket(index uint64) error {
+	_, err := csm.DelState(
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(bucketKey(index)))
+	return err
+}
+
+func (csm *candSM) putBucketAndIndex(bucket *VoteBucket) (uint64, error) {
+	index, err := csm.putBucket(bucket)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to put bucket")
+	}
+
+	if err := csm.PutVoterBucketIndex(bucket.Owner, index); err != nil {
```

Comment:
Calling public function inside of a private function is not recommended, but let's leave it to next PR.